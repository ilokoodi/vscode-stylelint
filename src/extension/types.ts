// eslint-disable-next-line node/no-unpublished-import
import type TypedEventEmitter from 'typed-emitter';

/**
 * VS Code extension event names.
 */
export interface ExtensionEvents {
	DidRegisterDocumentFormattingEditProvider: () => void;
}

/**
 * VS Code extension public API.
 */
export type PublicApi = TypedEventEmitter<ExtensionEvents> & {
	formattingReady: boolean;
};

/**
 * Extension API event names.
 */
export enum ApiEvent {
	DidRegisterDocumentFormattingEditProvider = 'DidRegisterDocumentFormattingEditProvider',
}
