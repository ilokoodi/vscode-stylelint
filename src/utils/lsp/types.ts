import type LSP from 'vscode-languageserver-protocol';

/**
 * Code actions for a particular rule.
 */
export interface RuleCodeActions {
	/**
	 * Action to disable the rule for a specific line.
	 */
	disableLine?: LSP.CodeAction;

	/**
	 * Action to disable the rule for an entire file.
	 */
	disableFile?: LSP.CodeAction;

	/**
	 * Action to show documentation for the rule.
	 */
	documentation?: LSP.CodeAction;
}
