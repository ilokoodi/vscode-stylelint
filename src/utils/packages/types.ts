// eslint-disable-next-line node/no-unpublished-import
import type stylelint from 'stylelint';

/**
 * Package manager identifiers.
 */
export type PackageManager = 'npm' | 'yarn' | 'pnpm';

/**
 * Options for resolving the Stylelint package.
 */
export type ResolverOptions = {
	packageManager?: PackageManager;
	stylelintPath?: string;
};

/**
 * Stylelint package resolution result.
 */
export type StylelintResolutionResult = {
	stylelint: stylelint.PublicApi;
	resolvedPath: string;
};

/**
 * A tracer function that can be used to log messages.
 */
export type TracerFn = (message: string, verbose?: string) => void;
