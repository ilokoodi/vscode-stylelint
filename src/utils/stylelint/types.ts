import type LSP from 'vscode-languageserver-protocol';
// eslint-disable-next-line node/no-unpublished-import
import type stylelint from 'stylelint';
import type { PackageManager } from '../packages';

/**
 * Diagnostics for a lint run.
 */
export type LintDiagnostics = {
	/**
	 * The diagnostics, each corresponding to a warning or error emitted by
	 * Stylelint.
	 */
	diagnostics: LSP.Diagnostic[];

	/**
	 * Raw output from Stylelint, if any.
	 */
	output?: string;
};

/**
 * Disable report rule names.
 */
export enum DisableReportRuleNames {
	Needless = '--report-needless-disables',
	InvalidScope = '--report-invalid-scope-disables',
	Descriptionless = '--report-descriptionless-disables',
	Illegal = 'reportDisables',
}

/**
 * Stylelint runner options.
 */
export type RunnerOptions = {
	config?: stylelint.Config | null;
	configBasedir?: string;
	configFile?: string;
	customSyntax?: string;
	ignoreDisables?: boolean;
	packageManager?: PackageManager;
	reportInvalidScopeDisables?: boolean;
	reportNeedlessDisables?: boolean;
	snippet?: string[];
	stylelintPath?: string;
	validate?: string[];
};

/**
 * Error thrown when a rule's option is invalid.
 */
export class InvalidOptionError extends Error {
	reasons: string[];

	constructor(warnings: { text: string }[]) {
		const reasons = warnings.map((warning) => warning.text);

		super(reasons.join('\n'));
		this.reasons = reasons;
	}
}

// TODO: Create type upstream
// https://github.com/stylelint/stylelint/pull/5696
/**
 * A Stylelint configuration error. Taken from
 * https://github.com/stylelint/stylelint/blob/551dcb5/lib/utils/configurationError.js
 */
export type ConfigurationError = Error & { code: 78 };
