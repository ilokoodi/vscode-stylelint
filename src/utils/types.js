'use strict';

const { CodeActionKind: VSCodeActionKind } = require('vscode-languageserver-types');

/**
 * Command IDs
 * @enum {string}
 */
const CommandId = {
	ApplyAutoFix: 'stylelint.applyAutoFix',
};

/**
 * Code action kinds
 * @enum {string}
 */
const CodeActionKind = {
	StylelintSourceFixAll: `${VSCodeActionKind.SourceFixAll}.stylelint`,
};

/**
 * Disable report rule names
 * @enum {string}
 */
const DisableReportRuleNames = {
	Needless: '--report-needless-disables',
	InvalidScope: '--report-invalid-scope-disables',
	Descriptionless: '--report-descriptionless-disables',
	Illegal: 'reportDisables',
};

/**
 * Error thrown when a rule's option is invalid.
 */
class InvalidOptionError extends Error {
	/** @param {{text: string}[]} warnings */
	constructor(warnings) {
		const reasons = warnings.map((warning) => warning.text);

		super(reasons.join('\n'));
		this.reasons = reasons;
	}
}

module.exports = {
	CommandId,
	CodeActionKind,
	DisableReportRuleNames,
	InvalidOptionError,
};
