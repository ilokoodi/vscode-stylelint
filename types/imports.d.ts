import * as postcss from 'postcss';
export import postcss = postcss;

import * as stylelint from 'stylelint';
export import stylelint = stylelint;

import * as vscode from 'vscode';
export import vscode = vscode;

import * as vscodeLanguageServer from 'vscode-languageserver';
export import vscodeLanguageServer = vscodeLanguageServer;

import * as vscodeLanguageServerTextDocument from 'vscode-languageserver-textdocument';
export import vscodeLanguageServerTextDocument = vscodeLanguageServerTextDocument;

import * as vscodeLanguageServerTypes from 'vscode-languageserver-types';
export import vscodeLanguageServerTypes = vscodeLanguageServerTypes;

export as namespace imports;
