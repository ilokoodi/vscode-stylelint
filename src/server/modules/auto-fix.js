'use strict';

const { WorkspaceChange } = require('vscode-languageserver-protocol');
const { CommandId } = require('../../utils/types');

/**
 * @implements {LanguageServerModule}
 */
class AutoFixModule {
	static id = 'auto-fix';

	/**
	 * The language server context.
	 * @type {LanguageServerContext}
	 */
	#context;

	/**
	 * The logger to use, if any.
	 * @type {winston.Logger | undefined}
	 */
	#logger;

	/**
	 * @param {LanguageServerModuleConstructorParameters} params
	 */
	constructor({ context, logger }) {
		this.#context = context;
		this.#logger = logger;
	}

	/**
	 * @param {lsp.TextDocument} document
	 * @returns {boolean}
	 */
	#shouldAutoFix(document) {
		return this.#context.options.validate.includes(document.languageId);
	}

	/**
	 * @returns {Partial<lsp.InitializeResult>}
	 */
	onInitialize() {
		return {
			capabilities: {
				executeCommandProvider: {
					commands: [CommandId.ApplyAutoFix],
				},
			},
		};
	}

	/**
	 * @returns {void}
	 */
	onDidRegisterHandlers() {
		this.#logger?.debug('Registering onExecuteCommand handler');

		this.#context.connection.onExecuteCommand(async ({ command, arguments: args }) => {
			this.#logger?.debug('Received onExecuteCommand', { command, arguments: args });

			if (command !== CommandId.ApplyAutoFix || !args) {
				return {};
			}

			/** @type { { version: number, uri: string } } */
			const identifier = args[0];
			const uri = identifier.uri;
			const document = this.#context.documents.get(uri);

			if (!document || !this.#shouldAutoFix(document)) {
				if (this.#logger?.isDebugEnabled()) {
					if (!document) {
						this.#logger.debug('Unknown document, ignoring', { uri });
					} else {
						this.#logger.debug('Document should not be auto-fixed, ignoring', {
							uri,
							language: document.languageId,
						});
					}
				}

				return {};
			}

			if (identifier.version !== document.version) {
				this.#logger?.debug('Document has been modified, ignoring', { uri });

				return {};
			}

			const workspaceChange = new WorkspaceChange();
			const textChange = workspaceChange.getTextEditChange(identifier);

			const edits = await this.#context.getFixes(document);

			edits.forEach((edit) => textChange.add(edit));

			this.#logger?.debug('Applying fixes', { uri, edits });

			try {
				const response = await this.#context.connection.workspace.applyEdit(workspaceChange.edit);

				if (!response.applied) {
					this.#logger?.debug('Failed to apply fixes', { uri, response });
				}
			} catch (error) {
				this.#logger?.debug('Failed to apply fixes', { uri, error });
			}

			return {};
		});

		this.#logger?.debug('onExecuteCommand handler registered');
	}
}

module.exports = {
	AutoFixModule,
};
