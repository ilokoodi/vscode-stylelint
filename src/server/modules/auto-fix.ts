import type { TextDocument } from 'vscode-languageserver-textdocument';
import { WorkspaceChange } from 'vscode-languageserver-protocol';
import type LSP from 'vscode-languageserver-protocol';
import { CommandId } from '../types';
import type {
	LanguageServerContext,
	LanguageServerModuleConstructorParameters,
	LanguageServerModule,
} from '../types';
import type winston from 'winston';

export class AutoFixModule implements LanguageServerModule {
	static id = 'auto-fix';

	/**
	 * The language server context.
	 */
	#context: LanguageServerContext;

	/**
	 * The logger to use, if any.
	 */
	#logger: winston.Logger | undefined;

	constructor({ context, logger }: LanguageServerModuleConstructorParameters) {
		this.#context = context;
		this.#logger = logger;
	}

	async #shouldAutoFix(document: TextDocument): Promise<boolean> {
		const options = await this.#context.getOptions(document.uri);

		return options.validate.includes(document.languageId);
	}

	onInitialize(): Partial<LSP.InitializeResult> {
		return {
			capabilities: {
				executeCommandProvider: {
					commands: [CommandId.ApplyAutoFix],
				},
			},
		};
	}

	onDidRegisterHandlers(): void {
		this.#logger?.debug('Registering onExecuteCommand handler');

		this.#context.connection.onExecuteCommand(async ({ command, arguments: args }) => {
			this.#logger?.debug('Received onExecuteCommand', { command, arguments: args });

			if (command !== CommandId.ApplyAutoFix || !args) {
				return {};
			}

			const identifier: { version: number; uri: string } = args[0];
			const uri = identifier.uri;
			const document = this.#context.documents.get(uri);

			if (!document || !(await this.#shouldAutoFix(document))) {
				if (this.#logger?.isDebugEnabled()) {
					if (!document) {
						this.#logger.debug('Unknown document, ignoring', { uri });
					} else {
						this.#logger.debug('Document should not be auto-fixed, ignoring', {
							uri,
							language: document.languageId,
						});
					}
				}

				return {};
			}

			if (identifier.version !== document.version) {
				this.#logger?.debug('Document has been modified, ignoring', { uri });

				return {};
			}

			const workspaceChange = new WorkspaceChange();
			const textChange = workspaceChange.getTextEditChange(identifier);

			const edits = await this.#context.getFixes(document);

			edits.forEach((edit) => textChange.add(edit));

			this.#logger?.debug('Applying fixes', { uri, edits });

			try {
				const response = await this.#context.connection.workspace.applyEdit(workspaceChange.edit);

				if (!response.applied) {
					this.#logger?.debug('Failed to apply fixes', { uri, response });
				}
			} catch (error) {
				this.#logger?.debug('Failed to apply fixes', { uri, error });
			}

			return {};
		});

		this.#logger?.debug('onExecuteCommand handler registered');
	}
}
