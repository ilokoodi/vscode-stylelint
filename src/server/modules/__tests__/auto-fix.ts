import { Position, TextEdit } from 'vscode-languageserver-types';
import { CommandId } from '../../types';

import { AutoFixModule } from '../auto-fix';

const mockContext = serverMocks.getContext();
const mockLogger = serverMocks.getLogger();

describe('AutoFixModule', () => {
	beforeEach(() => {
		mockContext.__options.validate = [];
		jest.clearAllMocks();
	});

	test('should be constructable', () => {
		expect(() => new AutoFixModule({ context: mockContext.__typed() })).not.toThrow();
	});

	test('onInitialize should return results', () => {
		const module = new AutoFixModule({ context: mockContext.__typed() });

		expect(module.onInitialize()).toMatchSnapshot();
	});

	test('onDidRegisterHandlers should register an auto-fix command handler', () => {
		const module = new AutoFixModule({ context: mockContext.__typed() });

		module.onDidRegisterHandlers();

		expect(mockContext.commands.on).toHaveBeenCalledTimes(1);
		expect(mockContext.commands.on).toHaveBeenCalledWith(
			CommandId.ApplyAutoFix,
			expect.any(Function),
		);
	});

	test('should auto-fix documents', async () => {
		mockContext.documents.get.mockReturnValue({
			uri: 'foo',
			languageId: 'bar',
			version: 1,
		});
		mockContext.__options.validate = ['bar'];
		mockContext.getFixes.mockReturnValue([TextEdit.insert(Position.create(0, 0), 'text')]);
		mockContext.connection.workspace.applyEdit.mockResolvedValue({ applied: true });

		const module = new AutoFixModule({ context: mockContext.__typed(), logger: mockLogger });

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)[1];

		const result = await handler({
			command: CommandId.ApplyAutoFix,
			arguments: [{ uri: 'foo', version: 1 }],
		});

		expect(result).toStrictEqual({});
		expect(mockContext.getFixes).toHaveBeenCalledWith({
			languageId: 'bar',
			uri: 'foo',
			version: 1,
		});
		expect(mockContext.connection.workspace.applyEdit.mock.calls[0]).toMatchSnapshot();
	});

	test('with incorrect command, should not attempt to auto-fix', async () => {
		const module = new AutoFixModule({ context: mockContext.__typed() });

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)[1];

		const result = await handler({ command: 'foo' });

		expect(result).toStrictEqual({});
		expect(mockContext.getFixes).not.toHaveBeenCalled();
	});

	test('with no arguments, should not attempt to auto-fix', async () => {
		const module = new AutoFixModule({ context: mockContext.__typed() });

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)[1];

		const result = await handler({ command: CommandId.ApplyAutoFix });

		expect(result).toStrictEqual({});
		expect(mockContext.getFixes).not.toHaveBeenCalled();
	});

	test('if no matching document exists, should not attempt to auto-fix', async () => {
		mockContext.documents.get.mockReturnValue(undefined);

		const module = new AutoFixModule({ context: mockContext.__typed(), logger: mockLogger });

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)[1];

		const result = await handler({
			command: CommandId.ApplyAutoFix,
			arguments: [{ uri: 'foo' }],
		});

		expect(result).toStrictEqual({});
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Unknown document, ignoring', { uri: 'foo' });
	});

	test('if document language ID is not in options, should not attempt to auto-fix', async () => {
		mockContext.documents.get.mockReturnValue({
			uri: 'foo',
			languageId: 'bar',
		});
		mockContext.__options.validate = ['baz'];

		const module = new AutoFixModule({ context: mockContext.__typed(), logger: mockLogger });

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)[1];

		const result = await handler({
			command: CommandId.ApplyAutoFix,
			arguments: [{ uri: 'foo' }],
		});

		expect(result).toStrictEqual({});
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Document should not be auto-fixed, ignoring',
			{
				uri: 'foo',
				language: 'bar',
			},
		);
	});

	test('with no debug log level and no valid document, should not attempt to log reason', async () => {
		mockContext.documents.get.mockReturnValue({
			uri: 'foo',
			languageId: 'bar',
		});
		mockContext.__options.validate = ['baz'];
		mockLogger.isDebugEnabled.mockReturnValue(false);

		const module = new AutoFixModule({ context: mockContext.__typed(), logger: mockLogger });

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)[1];

		const handlerParams = {
			command: CommandId.ApplyAutoFix,
			arguments: [{ uri: 'foo' }],
		};

		const result = await handler(handlerParams);

		expect(result).toStrictEqual({});
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Registering onExecuteCommand handler');
	});

	test('if the document has been modified, should not attempt to auto-fix', async () => {
		mockContext.documents.get.mockReturnValue({
			uri: 'foo',
			languageId: 'bar',
			version: 2,
		});
		mockContext.__options.validate = ['bar'];

		const module = new AutoFixModule({ context: mockContext.__typed(), logger: mockLogger });

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)[1];

		const result = await handler({
			command: CommandId.ApplyAutoFix,
			arguments: [{ uri: 'foo', version: 1 }],
		});

		expect(result).toStrictEqual({});
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Document has been modified, ignoring', {
			uri: 'foo',
		});
	});

	test('if fixes fail to apply, should log the response', async () => {
		const response = { applied: false, failureReason: 'foo' };

		mockContext.documents.get.mockReturnValue({
			uri: 'foo',
			languageId: 'bar',
			version: 1,
		});
		mockContext.__options.validate = ['bar'];
		mockContext.getFixes.mockReturnValue([TextEdit.insert(Position.create(0, 0), 'text')]);
		mockContext.connection.workspace.applyEdit.mockResolvedValue(response);

		const module = new AutoFixModule({ context: mockContext.__typed(), logger: mockLogger });

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)[1];

		const result = await handler({
			command: CommandId.ApplyAutoFix,
			arguments: [{ uri: 'foo', version: 1 }],
		});

		expect(result).toStrictEqual({});
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Failed to apply fixes', {
			uri: 'foo',
			response,
		});
	});

	test('if an error occurs while applying fixes, should log it', async () => {
		const error = new Error('foo');

		mockContext.documents.get.mockReturnValue({
			uri: 'foo',
			languageId: 'bar',
			version: 1,
		});
		mockContext.__options.validate = ['bar'];
		mockContext.getFixes.mockReturnValue([TextEdit.insert(Position.create(0, 0), 'text')]);
		mockContext.connection.workspace.applyEdit.mockRejectedValue(error);

		const module = new AutoFixModule({ context: mockContext.__typed(), logger: mockLogger });

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)[1];

		const result = await handler({
			command: CommandId.ApplyAutoFix,
			arguments: [{ uri: 'foo', version: 1 }],
		});

		expect(result).toStrictEqual({});
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Failed to apply fixes', {
			uri: 'foo',
			error,
		});
	});
});
