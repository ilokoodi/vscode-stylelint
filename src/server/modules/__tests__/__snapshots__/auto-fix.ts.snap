// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoFixModule onInitialize should return results 1`] = `
Object {
  "capabilities": Object {
    "executeCommandProvider": Object {
      "commands": Array [
        "stylelint.applyAutoFix",
      ],
    },
  },
}
`;

exports[`AutoFixModule should auto-fix documents 1`] = `
Array [
  Object {
    "changeAnnotations": undefined,
    "documentChanges": Array [
      Object {
        "edits": Array [
          Object {
            "newText": "text",
            "range": Object {
              "end": Object {
                "character": 0,
                "line": 0,
              },
              "start": Object {
                "character": 0,
                "line": 0,
              },
            },
          },
        ],
        "textDocument": Object {
          "uri": "foo",
          "version": 1,
        },
      },
    ],
  },
]
`;
